### PR TITLE
prevent endless parsing on bidirectional models

### PR DESCRIPTION
**Pull request by SunlightBro**
fixes issue #118 Stack Overflow Error

### CODE REVIEW

File: packages/freezed/lib/src/freezed_generator.dart [hunk 1]
```diff
@@ -174,6 +175,8 @@ class FreezedGenerator extends ParserGenerator<_GlobalData, Data, Freezed> {
 
     return _computeElementDataCache.putIfAbsent(classElement, () {
       if (!typeChecker.hasAnnotationOf(classElement)) return null;
+      if (_parsedElementCheckList.contains(classElement)) return null;
```
Comment (rrousselGit) [inline packages/freezed/lib/src/freezed_generator.dart:178]: That "return null" sounds fishy.

Maybe we should replace the putIfAbsent with a cache[key] = whatever; instead. And do nothing in this situation (whereas the other line actually assign null)

File: packages/freezed/lib/src/freezed_generator.dart [hunk 2]
```diff
@@ -164,6 +164,7 @@ $runtimeType(
 @immutable
 class FreezedGenerator extends ParserGenerator<_GlobalData, Data, Freezed> {
   final _computeElementDataCache = <ClassElement, Data>{};
+  final _parsedElementCheckList = <ClassElement>[];
```
Comment (rrousselGit) [inline packages/freezed/lib/src/freezed_generator.dart:167]: Should be a Set

File: packages/freezed/lib/src/freezed_generator.dart [hunk 3]
```diff
@@ -182,7 +185,10 @@ class FreezedGenerator extends ParserGenerator<_GlobalData, Data, Freezed> {
   }
 
   @override
-  Data parseElement(_GlobalData globalData, Element rawElement) {
+  Data parseElement(_GlobalData globalData, Element rawElement, {bool init}) {
```
Comment (rrousselGit) [inline packages/freezed/lib/src/freezed_generator.dart:188]: I don't quite understand what that init flag does. Could you explain?

Comment (SunlightBro) [reply packages/freezed/lib/src/freezed_generator.dart:188]: `parseElement()` gets called recursively through `_parseConstructorsNeedsGeneration()` and `_computeElementDataFor()`. 
But I only want to clear this `_parsedElementCheckList`, when its called by `GeneratorForAnnotation.generate()`,

Comment (rrousselGit) [reply packages/freezed/lib/src/freezed_generator.dart:188]: Why clear the list to begin with?

File: packages/freezed/lib/src/freezed_generator.dart (hunk 1 above)
Comment (SunlightBro) [reply packages/freezed/lib/src/freezed_generator.dart]: so something like this ?
```dart    
if (!typeChecker.hasAnnotationOf(classElement) || _parsedElementCheckSet.contains(classElement)) {
  return null;
} else {
  _parsedElementCheckSet.add(classElement);
  final data = parseElement(
    _GlobalData(hasDiagnostics: false, hasJson: false),
    classElement,
  );
  _computeElementDataCache[classElement] = data;
  return data;
}
```

**Comment by rrousselGit** [comment]
Could you add a test where it's not a List<Appointment> and try to use deep copy on it?

We should be able to write:

person.copyWith.appointment.person.appointment.person(name: 'foo')

**Comment by SunlightBro** [comment]
Ok I added a test.

**Comment by rrousselGit** [comment]
Thanks!

**Pull request merged**